crates/multi-transaction-payment/src/tests.rs

```diff
@@ -0,0 +1,417 @@
+            assert!(dry_run_native(6, Some(3), err, Pays::Yes) < baseline);
+
+            // weight of 0: still pays for extrinsic overhead
+            assert!(dry_run_native(6, Some(0), err, Pays::Yes) > 0);
+
+            // Pays::No completely refunds everything regardless of weight
+            assert_eq!(dry_run_native(6, None, err, Pays::No), 0);
+            assert_eq!(dry_run_native(6, Some(6), err, Pays::No), 0);
+            assert_eq!(dry_run_native(6, Some(3), err, Pays::No), 0);
+            assert_eq!(dry_run_native(6, Some(0), err, Pays::No), 0);
+        }
+    });
+}
+
+#[test]
+/// Like `pay_in_native_fee_refunds`, but using a non-native fee currency
+fn pay_in_swapped_fee_refunds() {
+    new_test_ext().execute_with(|| {
+        let baseline = dry_run_swapped(6, None, false, Pays::Yes);
+        assert!(baseline.fee_cost > 0); // decrease in fee currency, no change in native
+        assert_eq!(baseline.native_cost, 0);
+
+        // `Pays::No` completely refunds everything regardless of weight or err
+        let refunded_baseline = dry_run_swapped(6, None, false, Pays::No);
+        assert_eq!(refunded_baseline.fee_cost, baseline.fee_cost);
+        assert!(refunded_baseline.native_cost < 0); // refunded in native currency
+        assert_eq!(dry_run_swapped(6, Some(6), false, Pays::No), refunded_baseline);
+        assert_eq!(dry_run_swapped(6, Some(3), false, Pays::No), refunded_baseline);
+        assert_eq!(dry_run_swapped(6, Some(0), false, Pays::No), refunded_baseline);
+        assert_eq!(dry_run_swapped(6, None, true, Pays::No), refunded_baseline);
+        assert_eq!(dry_run_swapped(6, Some(6), true, Pays::No), refunded_baseline);
+        assert_eq!(dry_run_swapped(6, Some(3), true, Pays::No), refunded_baseline);
+        assert_eq!(dry_run_swapped(6, Some(0), true, Pays::No), refunded_baseline);
+
+        for err in [true, false] {
+            assert_eq!(dry_run_swapped(6, Some(6), err, Pays::Yes), baseline);
+
+            // Partial refund because less weight was used
+            let cost = dry_run_swapped(6, Some(3), err, Pays::Yes);
+            assert_eq!(cost.fee_cost, baseline.fee_cost);
+            assert!(cost.native_cost < 0); // refunded in native currency
+            assert!(cost.native_cost > refunded_baseline.native_cost); // not refunded as much as when its free
+        }
+    });
+}
+
+#[test]
+/// test that tips are paid in the fee currency, not the native currency
+fn pay_in_swapped_fee_with_tip() {
+    new_test_ext().execute_with(|| {
+        let baseline = dry_run_swapped_with_tip(6, None, false, Pays::Yes, 0);
+        assert!(baseline.fee_cost > 0); // decrease in fee currency, no change in native
+        assert_eq!(baseline.native_cost, 0);
+
+        // Paying tip increases the amount of fee taken in the fee currency
+        let cost = dry_run_swapped_with_tip(6, None, false, Pays::Yes, 1);
+        assert!(cost.fee_cost > baseline.fee_cost);
```

Comment:
Is there a way to know the expected amount resulting from a conversion, including slippage? These inequality assertions seem rather weak